libraries/Microsoft.Bot.Builder.Ai/LocaleConverter.cs

```diff
@@ -0,0 +1,200 @@
+            };
+            foreach (string locale in new string[] { "en-za", "en-ie", "en-gb", "en-ca", "fr-ca", "zh-cn", "zh-sg", "zh-hk", "zh-mo", "zh-tw" })
+            {
+                mapLocaleToFunction[locale] = yearMonthDay;
+            }
+            foreach (string locale in new string[] { "en-au", "fr-be", "fr-ch", "fr-fr", "fr-lu", "fr-mc", "de-at", "de-ch", "de-de", "de-lu", "de-li" })
+            {
+                mapLocaleToFunction[locale] = dayMonthYear;
+            }
+            mapLocaleToFunction["en-us"] = monthDayYEar;
+        }
+
+        //check if a specific locale is available
+        public bool IsLocaleAvailable(string locale)
+        {
+            return mapLocaleToFunction.ContainsKey(locale);
+        }
+
+        //extract date and time from a sentence using Microsoft Recognizer
+        private List<TextAndDateTime> extractDate(string message, string fromLocale)
+        {
+            List<TextAndDateTime> fndDates = new List<TextAndDateTime>();
+            var culture = Culture.English; 
+            if (fromLocale.StartsWith("fr"))
+            {
+                culture = Culture.French;
+            }
+            else if (fromLocale.StartsWith("de"))
+            {
+                culture = Culture.German;
+            }
+            else if (fromLocale.StartsWith("pt"))
+            {
+                culture = Culture.Portuguese;
+            }
+            else if (fromLocale.StartsWith("zh"))
+            {
+                culture = Culture.Chinese;
+            }
+            else if (fromLocale.StartsWith("es"))
+            {
+                culture = Culture.Spanish;
+            }
+            var model = DateTimeRecognizer.GetInstance().GetDateTimeModel(culture);
+            var results = model.Parse(message);
+            foreach (ModelResult result in results)
+            {
+                var resolutionValues = (IList<Dictionary<string, string>>)result.Resolution["values"];
+                string type = result.TypeName.Split('.').Last();
+                DateTime moment = new DateTime();
+                if (type.Contains("date") && !type.Contains("range"))
+                {
+                    moment = resolutionValues.Select(v => DateTime.Parse(v["value"])).FirstOrDefault();
+
+                }
+                else if (type.Contains("date") && type.Contains("range"))
+                {
+                    moment = DateTime.Parse(resolutionValues.First()["start"]);
+                }
+                else
+                    continue;
+                var curDateTimeText = new TextAndDateTime
+                {
+                    DateTimeObj = moment,
+                    Text = result.Text
+                };
+                fndDates.Add(curDateTimeText);
+            }
+            return fndDates;
+        }
+
+        //Convert a message from locale to another locale
+        public Task<string> Convert(string message, string fromLocale, string toLocale)
```

Comment:
>Task [](start = 15, length = 4)

might be nicer to use "async Task" otherwise you're leaking exceptions out of the Task "container"

Reply:
Updated to be async Task that would be better